pkg/utils/kubernetes/highavailability.go

```diff
@@ -60,14 +60,33 @@ func GetTopologySpreadConstraints(
 	}
 
 	var (
+		// Enforcing a spread over zones is required when there are:
+		// - multiple zones
+		// - AND
+		// - the failure tolerance type is 'nil' (seed/shoot system component case) or 'zone' (shoot control-plane case)
+		zoneSpreadRequired = numberOfZones > 1 && (failureToleranceType == nil || *failureToleranceType == gardencorev1beta1.FailureToleranceTypeZone)
+
+		// Enforcing a spread over hosts is required when:
+		// - it is explicitly requested via 'enforceSpreadAcrossHosts' argument
+		// - OR
+		// - a failure tolerance type is set
+		hostSpreadRequired = enforceSpreadAcrossHosts || ptr.Deref(failureToleranceType, "") != ""
+
 		minDomainsHosts   *int32
-		whenUnsatisfiable = corev1.ScheduleAnyway
+		whenUnsatisfiable corev1.UnsatisfiableConstraintAction
 	)
-	if ptr.Deref(failureToleranceType, "") != "" || enforceSpreadAcrossHosts {
+
+	if hostSpreadRequired {
 		whenUnsatisfiable = corev1.DoNotSchedule
 		minDomainsHosts = calculateMinDomains(3, maxReplicas)
 	}
 
+	// Zone and host spreads in combination can only work when the number of replicas don't exceed twice the number of zones.
+	// A resulting deadlock must be prevented by soften the host constraint to 'ScheduleAnyway', see https://github.com/kubernetes/kubernetes/issues/109364.
+	if !hostSpreadRequired || (zoneSpreadRequired && maxReplicas >= 2*numberOfZones) {
+		whenUnsatisfiable = corev1.ScheduleAnyway
+	}
```

Comment:
I am not sure I understand this pull request to its fullest extent, but would this not always mean for `istio` that the pods get `ScheduleAnyway` as their HPA resources have higher `maxReplicas` than twice the zone count? My guess is that we do not want `istio` pods running on the same node, but rather always spread them.
WDYT?

Reply:
Actually, you got me thinking about this workaround using `maxSkew: 2` (earlier) or `ScheduleAnyway` (with this PR).
The assumption was that either or both are needed to prevent the deadlock described in https://github.com/kubernetes/kubernetes/issues/109364, but with `ScheduleAnyway` having better balance which is however problematic for new seed clusters, because those are usually not equipped with the amount of nodes to reach the desired balance with `ScheduleAnyway` - on best effort basis.

Returning to https://github.com/kubernetes/kubernetes/issues/109364:
This deadlock happens if nodes across zones are imbalanced, e.g.:

Z1 - 3 nodes
Z2 - 1 node
Z3 - 1 node

This node imbalance **and** the number of replicas decide when a potential deadlock happens. It turned out in my tests, **the today used formula `maxReplicas >= 2*numZones` doesn't cover any of such scenarios anyway**.
For instance, a deployment with 5 replicas can't be scheduled with a node distribution as shown above and the following constraints:

```
      topologySpreadConstraints:
      - maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: DoNotSchedule
        labelSelector:
          matchLabels:
            app: pause
        matchLabelKeys:
        - pod-template-hash
      - maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: DoNotSchedule
        labelSelector:
          matchLabels:
            app: pause
        matchLabelKeys:
        - pod-template-hash
```

Luckily, the `cluster-autoscaler` helps to get out of this situation by scaling up the necessary node group. My conclusion is that it's safe to get rid of this `maxSkew: 2` or `ScheduleAnyway` workaround and instead rely on the CA.

Does that make sense?
/cc @ialidzhikov

Reply:
I wanted to look deeper into https://github.com/kubernetes/kubernetes/issues/109364 and understand it. And have idea whether and how it affects our use cases. 

The `cluster-autoscaler` argument seems strong enough for me. If adding a new Node in a zone helps avoiding https://github.com/kubernetes/kubernetes/issues/109364 then cluster-autoscaler should be able to add a new Node and we should be safe.

Reply:
My primary concern was that this block kind of reverts what I tried to achieve with #8970. For istio, we had the issue that in small seed cluster we may not have enough nodes to force a spread. However, if a seed is bigger there are enough nodes so that a spread should be enforced. #8970 implements this specifically for istio. It at least seems like the `if` block would undo parts of it.

Reply:
> The cluster-autoscaler argument seems strong enough for me. If adding a new Node in a zone helps avoiding https://github.com/kubernetes/kubernetes/issues/109364 then cluster-autoscaler should be able to add a new Node and we should be safe.

Yes, I tested this for several scenarios successfully. I will close this PR and open a new one for the said reasons.

Reply:
Thank you, @timuthy !